### ISSUE TITLE
How to specify different user to execute sql in yarn per-job mode

### ISSUE THREAD

**Issue by wooplevip** [ververica/flink-sql-gateway#36]
Hi all,

We can set HADOOP_USER_NAME for each job when to submit  job in yarn per-job mode via flink command. But how to specify different user for each job via REST API in yarn per-job mode? Thanks.

**Comment by godfreyhe** [comment]
which flink command supports "HADOOP_USER_NAME" ?

do you mean `--yarnname` ?

**Comment by wooplevip** [comment]
Hi @godfreyhe ,

Maybe it is not correct expression for `set HADOOP_USER_NAME`. 
For example, we add `HADOOP_USER_NAME` to `optsArray` and cmd is `${FLINK_HOME}/bin/flink run ...` then  invoke `Process.apply(cmd, None, optsArray: _*).!`.

**Comment by godfreyhe** [comment]
can you give a completed flink run command ?

**Comment by wooplevip** [comment]
The command is like 
`flink-1.10.0/bin/flink run --jobmanager yarn-cluster --detached --class com.xxx.StreamApp --parallelism 1 --yarnname test --yarnjobManagerMemory 1024 --yarnqueue default --yarnslots 1 --yarntaskManagerMemory 1024 -yD flink.master=yarn-cluster -yD group.id=test1586745136882`

`Process.apply(cmd, None, optsArray: _*).!` is equivalent to run `export HADOOP_USER_NAME=user1` firstly then run  above command.

**Comment by godfreyhe** [comment]
so `user1` is not from `flink run` command. you run `export HADOOP_USER_NAME=user1` just before `flink run` ?

**Comment by wooplevip** [comment]
Yes. As I understand it, we have to set `HADOOP_USER_NAME` to system environment for each different user. So I did a further testing. I set `HADOOP_USER_NAME` to system environment forcedly before submit a select sql job refer to  [this blog](https://blog.sebastian-daschner.com/entries/changing_env_java). It works as my expectation. But maybe it is not common way for this issue. Any other better suggestion? Thanks.

**Comment by godfreyhe** [comment]
have you ever tried adding "HADOOP_USER_NAME" into `env.java.opts.taskmanager` option which is `flink-conf.yaml` ?

**Comment by wooplevip** [comment]
It does not work for yarn-per job mode. `HADOOP_USER_NAME` must be in gateway process system environment from my testing.

**Comment by wooplevip** [comment]
If I only add `System.setProperty("HADOOP_USER_NAME", "user1");` before `final ProgramDeployer deployer = new ProgramDeployer(configuration, jobName, pipeline);`, it also works. I will do more testing to double check.

**Comment by fpompermaier** [comment]
Why bot copying what Livy does for spark and add a doAs() service to the REST APIs?

**Comment by wooplevip** [comment]
@fpompermaier thanks, good idea, I will look into Livy implement.

@godfreyhe as workaround, I can invoke `UserGroupInformation.setLoginUser(null);`  and `System.setProperty("HADOOP_USER_NAME", "user1");` before execute job operation.

Could you think about this requirement? Thanks very much.

**Comment by godfreyhe** [comment]
@wooplevip, maybe @fpompermaier 's suggestion is good approach.

**Comment by wooplevip** [comment]
Thanks. Do you have any plan to support multiple users can share the same gateway server?

**Comment by godfreyhe** [comment]
you can take this ticket if you have time.

**Comment by wooplevip** [comment]
OK, I will have a try.

**Comment by godfreyhe** [comment]
@wooplevip could you briefly describe your design before submitting a pr ?

**Comment by wooplevip** [comment]
Hi @godfreyhe ,

There are two scenarios that we should support: insecure and secure cluster. Please help to review the following modification, thanks.

1. Modify gateway startup procedure

![startup _3_.png](https://i.loli.net/2020/04/28/LtTdUKGhO2ebcqv.png)

2. Add proxyUser and doAs user for REST API

![image.png](https://i.loli.net/2020/04/28/eqLKouBbZ8UhHD5.png)

**Comment by wooplevip** [comment]
Hi @godfreyhe ,

Can we add a `Map<String, String> executionConf` field in `StatementExecuteRequestBody` to get properties user specified, for example, `proxyUser`,`queue`, `owner` and so on ?
